### ISSUE TITLE
iptables: Issues with additional chains (iptables -N)

### ISSUE THREAD

**Issue by primeos** [NixOS/nixpkgs#21737]
## Issue description

The [`services.networking.firewall`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/firewall.nix) module only flushes and deletes the following chains: "nixos-fw nixos-fw-accept nixos-fw-log-refuse nixos-fw-refuse FW_REFUSE" (I don't know why FW_REFUSE is in there btw) using the following code:
```
    # Flush the old firewall rules.  !!! Ideally, updating the
    # firewall would be atomic.  Apparently that's possible
    # with iptables-restore.
    ip46tables -D INPUT -j nixos-fw 2> /dev/null || true
    for chain in nixos-fw nixos-fw-accept nixos-fw-log-refuse nixos-fw-refuse FW_REFUSE; do
      ip46tables -F "$chain" 2> /dev/null || true
      ip46tables -X "$chain" 2> /dev/null || true
    done
```

This causes problems when adding new chains (e.g. "networking.firewall.extraCommands = "iptables -N demo-fw") since this fails after running `nixos-rebuild switch` for the second time (and making changes so that the firewall derivation will be build again). Also the additional chains won't be flushed which results in these chains getting duplicate rules.

I would suggest solving this by flushing all chains followed by deleting all chains (they must be empty before we can delete them). This can be done by simply running the following two commands:
```bash
ip46tables -F
ip46tables -X
```

If there are no objections I could also create a PR for that :smile:

### Steps to reproduce

Do the following:
1) Add `networking.firewall.extraCommands = "iptables -N demo-fw"` to your `configuration.nix`
2) Run `nixos-rebuild test` (or switch)
3) Make some changes to `networking.firewall.extraCommands` e.g.:
```
iptables -N demo-fw
iptables -I demo-fw -p tcp --dport 22 -i eth0 -j DROP
```
4) Re-run `nixos-rebuild test` (or switch)

## Technical details

* System: (NixOS: `16.09.1368.435b5f8` (but this affects all recent versions up to master))

PS: I really hope I haven't missed a valid reason for this... :D

cc @edolstra @wkennington

**Comment by jokogr** [comment]
You can use `networking.firewall.extraStopCommands` if this is an issue for you, e.g.:

```
networking.firewall.extraStopCommands = ''
  iptables -F demo-fw
  iptables -X demo-fw
'';
```

**Comment by primeos** [comment]
@jokogr Yes, I could use that option (or reset the complete firewall first with `networking.firewall.extraCommands` to get rid of the predefined rules, which is another issue imho), however that's more like a dirty hack than a real solution for me - I. e. this doesn't work for me (#33C3).

But thanks for sharing that :smile:

I guess opening an issue for this wasn't the best idea in the first place. I'll have a closer look at the module and make a PR within the next few days, it's probably better to discuss specific changes on a PR.

**Comment by jokogr** [comment]
@primeos no, please keep the issue open a bit more, I also like to see what `FW_REFUSE` is about 👍

**Comment by primeos** [comment]
commit e4051e105c2dc677e7a4650e0c8a9d4053b3b836
```
# The "FW_REFUSE" chain performs logging and
# rejecting/dropping of packets.
ip46tables -N FW_REFUSE
```

**Comment by jokogr** [comment]
As such, let's remove it, let me know if you want me to do a PR for that

**Comment by primeos** [comment]
@jokogr - I figured that change will probably be way more complex than I initially thought... The referenced PR (above) theoretically contains a preview but I think it'll break quite some stuff.

I noticed some other things that I'll probably submit first but I'll try to remember to later reference the final PR here :smile:
